content/gocourse-2019-05-melbourne.slide

```diff
@@ -910,36 +910,45 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
 - Add locking for proper use of sync.Map
 - Bonus points: Add a third `Storer` implementation using [[https://github.com/syndtr/goleveldb][LevelDB]]
 
+* Channels
 
-* Concurrency (1/2)
+- Communication mechanism (conduit) that lets goroutines exchange data
+- Each channel allows goroutines to send/receive a particular data type (called element type)
+- Declare channel using `chan` keyword
+- Create unbuffered channel
+	ch := make(chan int) // ch is an unbuffered channel of `int` element type
+- Create buffered channel
+	ch := make(chan int, 10) // ch is a buffered channel of size 10 and type `int`
+- Close channel using `close()` function
 
-"Don't communicate by sharing memory, share memory by communicating."
-"Concurrency is not parallelism."
-"Concurrency is about structure, parallelism is about execution."
- - [[https://go-proverbs.github.io/][Rob Pike]] ([[https://www.youtube.com/watch?v=cN_DpYBzKso][Talk]], [[https://talks.golang.org/2012/waza.slide][Slides]])
+* Channel Operations
 
-*Overview:* *built-in* *language* *features*
-Goroutines (`go`)
-Channels (`chan`, `<-`)
-Select (`select`)
+- Send and Receive values with the channel operator <-
 
+	ch <- v // send v to the channel ch
+	v := <-ch // receive from ch and assign it to v
 
-* Concurrency (2/2)
+- Send to buffered channel blocks only when the channel is full
+- Sender can close the channel to indicate no more values will be sent
+- Receiver can test whether a channel has been closed
 
-Concurrency isn't easy - we are only scratching the surface.
+	v, ok := <-ch // ok is false if channel is closed and no more data is available
 
-- [[https://goplay.space/#YtuJgIuIswH][Goroutines]] ([[https://play.golang.org/p/YtuJgIuIswH][Playground]]`*`)
-- [[https://goplay.space/#5M0f3-YY44W][Channels]] ([[https://play.golang.org/p/5M0f3-YY44W][Playground]]`*`)
-- [[https://goplay.space/#FEWHmUA0QtN][Buffered Channels]]
-- [[https://goplay.space/#ANO8rpfNynY][Range and close]] ([[https://play.golang.org/p/ANO8rpfNynY][Playground]]`*`)
-- [[https://goplay.space/#R1qgjOsPSGk][Select]] ([[https://play.golang.org/p/R1qgjOsPSGk][Playground]]`*`)
-- [[https://goplay.space/#4TGM29YbvX4][Deadlocks]] ([[https://play.golang.org/p/4TGM29YbvX4][Playground]]`*`)
+- Sending message on a closed/nil channel causes the program to panic
```

Comment:
Add another point:

* Receive on a closed channel never blocks

At this point it may make sense to have one slide for sending and another for receiving/range, as this slide is quite full now. This obviously goes against our earlier advice of having the code samples do a send and receive in one sample, but I think for presenting the material, it wont hurt. Otherwise perhaps move range to the next slide.

Reply:
I've added the point you've mentioned. That's good it makes all statements symmetrical. I felt your idea of having send and receive next to each other is more easy to read, understand and remember. So I've actually moved just the range to the next slide while others are on the same slide but they seem to look okay together. Please have a look. Would you still think I need to split send and receive separately for the slides ?

Reply:
Send and receive on one slide is fine if it all fits (and it does seem to), so this is fine.